### PR TITLE
[FLINK-1980] allowing users to decorate input streams

### PR DESCRIPTION
**Pull request by sekruse**
* add a decorateInputStream() method as hook in FileInputFormat
* provide a InputStreamFSInputWrapper to conveniently wrap InputStreams
* base existing .deflate file support on these changes
* add a test to verify the decoration

### CODE REVIEW

**Comment by rmetzger** [comment]
Thank you for this pull request. I'll review it now.

The first thing I saw is that the build has failed:
```
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 10.728 s
[INFO] Finished at: 2015-05-07T12:08:03+00:00
[INFO] Final Memory: 25M/448M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.rat:apache-rat-plugin:0.11:check (default) on project flink-parent: Too many files with unapproved license: 1 See RAT report in: /home/travis/build/apache/flink/target/rat.txt -> [Help 1]
``` 
https://travis-ci.org/apache/flink/jobs/61608003

You can also do a local `mvn clean install -DskipTests` to run the rat plugin.

File: flink-core/src/main/java/org/apache/flink/api/common/io/InputStreamFSInputWrapper.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
+package org.apache.flink.api.common.io;
```
Comment (rmetzger) [inline flink-core/src/main/java/org/apache/flink/api/common/io/InputStreamFSInputWrapper.java]: Missing license header, that's also why the build fails ;)

File: flink-core/src/main/java/org/apache/flink/api/common/io/InputStreamFSInputWrapper.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        int read = inStream.read();
+        if (read != -1) {
+            this.pos++;
+        }
+        return read;
+    }
+
+    @Override
+    public int read(byte[] b, int off, int len) throws IOException {
+        int numReadBytes = inStream.read(b, off, len);
```
Comment (rmetzger) [inline flink-core/src/main/java/org/apache/flink/api/common/io/InputStreamFSInputWrapper.java]: This method will return `-1` when the stream is over. Will this corrupt the `pos` ? (Above, you're checking if `read != -1` ?)

File: flink-core/src/test/java/org/apache/flink/api/common/io/FileInputFormatTest.java [hunk 3]
```diff
@@ -18,24 +18,20 @@
 
 package org.apache.flink.api.common.io;
 
-import java.io.BufferedOutputStream;
-import java.io.BufferedWriter;
-import java.io.File;
-import java.io.FileOutputStream;
-import java.io.FileWriter;
-import java.io.IOException;
+import java.io.*;
```
Comment (rmetzger) [inline flink-core/src/test/java/org/apache/flink/api/common/io/FileInputFormatTest.java]: As per our coding guidelines, star imports are not allowed: "Do not use wildcard imports in the core files. They can cause problems when adding to the code and in some cases even during refactoring. Exceptions are the Tuple classes, Tuple-related utilities, and Flink user programs, when importing operators/functions. Tests are a special case of the user programs." http://flink.apache.org/coding_guidelines.html

**Comment by rmetzger** [comment]
The change looks good, its not completely consistent with our coding guidelines. But a few runs of "mvn install -DskipTests" will resolve this ;)
(also, the indentation is done using spaces, we use tabs).

**Comment by sekruse** [comment]
Alright, I incorporated all the feedback. The build seems to be passing now.

**Comment by rmetzger** [comment]
+1 to merge.

**Comment by StephanEwen** [comment]
Looks good. One comment, though:
 - The position tracking in the stream wrapper is a nice idea, but since the `skip()` method is anyways only a hint (it may skip less if it wants) and seeking works only in one direction, why not skip seek support completely? Also simplifies the code and makes it even more lightweight.

As a followup: Should we add more built-in decompressors for other file endings, like `*.gz` ?

**Comment by rmetzger** [comment]
I think Sebastian has already filed a JIRA for adding gz read support.

**Comment by sekruse** [comment]
I wanted to integrate the gz support with the decorateStream method, therefore I was waiting for this PR to be merged.
I can of course skip the seek, but I thought that forward seeking would be useful as it enables splittable file formats.

**Comment by StephanEwen** [comment]
That is a good point. It is a bit tricky, though, since the skip method does not guarantee efficient skipping. It also reserves the right to skip less (and return the number of bytes skipped, iirc).

**Comment by sekruse** [comment]
I see your point and double-checked it with the Java doc. I will adjust the seek method accordingly.

**Comment by rmetzger** [comment]
I think the PR is good to merge.

**Comment by StephanEwen** [comment]
Looks good.

Will merge this in my next batch...

**Pull request closed without merging**